procrastinate/blueprints.py

```diff
@@ -0,0 +1,137 @@
+import functools
+from typing import TYPE_CHECKING, Any, Callable, List, Optional
+
+from procrastinate import jobs, protocols, retry
+
+if TYPE_CHECKING:
+    from procrastinate import app, tasks
+
+
+class Blueprint(protocols.TaskCreator):
+    """
+    A Blueprint provides a way to declare tasks that can be registered on an
+    :class:`~procrastinate.App` later::
+
+        bp = Blueprint()
+
+        ... declare tasks ...
+
+        app.register(bp)
+
+    Notes
+    -----
+    Calling a blueprint task before the it is bound to an app will raise an
+    Assertion error::
```

Comment:
Most (if not all) of the exceptions we get from interacting with Procrastinate inherit our main exception class. I think it's worth creating a new custom exception type (and add it to `docs/reference.py`)